### PR TITLE
Add OAuth2.0 and accessing Books API with access token

### PR DESCRIPTION
**Pull request by abbymapes**
OAuthCallbackServlet.java and OAuthServlet.java
- Added two OAuth servlets to retrieve Credential from OAuth server when a user logs in
- Follows AuthorizationCodeFlow to retrieve and store a Credential for users in datastore

OAuthHelper.java
- Helper methods to initialize AuthorizationCodeFlow and override methods necessary for OAuth serlvets
- Methods to check and refresh user's credentials once the access token has expired

BookUtils.java
- Added methods to access MyLibrary requests from Google Books API with authenticated user's access token

BookAgent.java
- Added handling for books.library requests, when users request to see books in their specific bookshelves
- This table works like book search tables, so added handling to access next, previous, description, and preview functionality for mylibrary vs. generic search request

Frontend
- Added preliminary frontend display for bookshelf requests, in order to ensure outputs and intents are functioning correctly

Updated some classes and servlets to account for the new books.library intent handling, and differentiating between authenticated and non-authenticated users

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/agents/BooksAgent.java [hunk 1]
```diff
@@ -125,35 +126,48 @@ public void setParameters(Map<String, Value> parameters)
...
           // Store Book results and new indices
           BooksMemoryUtils.storeBooks(bookResults, startIndex, sessionID, queryID, datastore);
           replaceIndices(sessionID, queryID);
+          this.output = "Here's the next page of results.";
         }
       }
       setBookListDisplay();
       this.redirect = queryID;
-      this.output = "Here's the next page of results.";
+      // this.output = "Here's the next page of results.";
```
Comment (aliu917) [inline portfolio/src/main/java/com/google/sps/agents/BooksAgent.java:164]: remove comments

File: portfolio/src/main/java/com/google/sps/agents/BooksAgent.java [hunk 2]
```diff
@@ -274,6 +347,19 @@ private String bookToString(Book book) {
...
 
+  private ArrayList<String> allLowerCaseList(ArrayList<String> list) {
+    ArrayList<String> lowerCaseList = new ArrayList<String>();
+    for (String word : list) {
+      lowerCaseList.add(word.toLowerCase());
+    }
+    return lowerCaseList;
+  }
+
+  private String listToString(ArrayList<String> list) {
```
Comment (aliu917) [inline portfolio/src/main/java/com/google/sps/agents/BooksAgent.java:358]: If you're just doing gson.toJson(object), then it doesn't really matter what the type is as the function takes in an Object so you don't need both listToString and bookListToString (rather, just make the input parameter a generic arraylist and call the same function).

File: portfolio/src/main/java/com/google/sps/data/BookQuery.java [hunk 3]
```diff
@@ -38,30 +38,57 @@
...
    *
    * @param String userInput
    * @param parameters parameter Map from Dialogflow
    * @return BookQuery object
    */
   public static BookQuery createBookQuery(String userInput, Map<String, Value> parameters)
       throws IllegalArgumentException {
+    return createBookQuery(userInput, parameters, false);
+    /*
     if (userInput == null || userInput.isEmpty()) {
```
Comment (aliu917) [inline portfolio/src/main/java/com/google/sps/data/BookQuery.java:66]: commented code

File: portfolio/src/main/java/com/google/sps/data/Indices.java [hunk 4]
```diff
@@ -86,4 +86,11 @@ public boolean getHasMore() {
   public boolean getHasPrev() {
     return this.hasPrev;
   }
+
+  private int getTotalPages(int totalResults, int displayNum) {
+    if (totalResults % displayNum == 0) {
```
Comment (aliu917) [inline portfolio/src/main/java/com/google/sps/data/Indices.java:91]: Would doing Math.ceil(totalResults / displayNum) be the same as this?

File: portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java [hunk 5]
```diff
@@ -46,6 +47,21 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     response.getWriter().write(json);
   }
 
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response)
+      throws IOException, ServletException {
+    System.out.println("Got to auth servlet");
+    if (userService.isUserLoggedIn()) {
+      // response.sendRedirect(url);
```
Comment (aliu917) [inline portfolio/src/main/java/com/google/sps/servlets/AuthServlet.java:55]: commented code

File: portfolio/src/main/java/com/google/sps/servlets/OAuthCallbackServlet.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+@WebServlet("/oauth2callback")
+public class OAuthCallbackServlet extends AbstractAuthorizationCodeCallbackServlet {
+  UserService userService = UserServiceFactory.getUserService();
+  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  @Override
+  protected void onSuccess(HttpServletRequest req, HttpServletResponse resp, Credential credential)
+      throws ServletException, IOException {
+    System.out.println("Success callback servlet");
```
Comment (aliu917) [inline portfolio/src/main/java/com/google/sps/servlets/OAuthCallbackServlet.java:26]: Can we switch all system print statements to logs? That way, if we ever run into this with testing, it won't cause nullpointers.

File: portfolio/pom.xml [hunk 7]
```diff
@@ -196,7 +214,7 @@
         <version>2.19.1</version>
         <configuration>
           <redirectTestOutputToFile>true</redirectTestOutputToFile>
-          <!-- <skipTests>true</skipTests> -->
+          <skipTests>true</skipTests>
```
Comment (mihirapatel) [inline portfolio/pom.xml:217]: keep commented out for github

**Pull request closed without merging**